compaction/compaction_strategy_state.hh

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "time_window_compaction_strategy.hh"
+#include "leveled_compaction_strategy.hh"
+#include <utility>
+#include <type_traits>
+#include <variant>
+
+namespace compaction {
+
+using namespace sstables;
+
+class compaction_strategy_state {
+public:
+    struct default_empty_state {};
+    using states_variant = std::variant<default_empty_state, leveled_compaction_strategy_state, time_window_compaction_strategy_state>;
+private:
+    states_variant _state;
+public:
+    explicit compaction_strategy_state(states_variant state) : _state(std::move(state)) {
+    }
+
+    template <typename StateType>
+    requires requires (StateType st) {
+        { states_variant{} = st };
+    }
+    StateType& get() noexcept {
+        return std::get<StateType>(_state);
+    }
+
+    static compaction_strategy_state make(const compaction_strategy& cs);
+};
```

Comment:
I think a void* is better as the state type here. It's not a true variant, it can only hold one type at a time.

We can have something like

```c++
template <typename State>
class compaction_strategy_impl_with_state : public compaction_strategy_impl {
protected:
    compaction_strategy_state make_state() const { return static_cast<State*>(compaction_strategy_state::from_ptr(new State())); }
    void destroy_state(compaction_strategy_state s) const {  delete static_cast<State*>(s.to_ptr()); }
    State& state(compaction_strategy_state s) const { 
          return *static_cast<State*>(s.to_ptr());
    }
}

Reply:
Or we can have a pair of both the compaction_strategy_impl* and a void*, to make it self contained for destruction.

Reply:
Not sure it's worth the effort. It's just a small wart and we aren't adding new strategies every thursday.

Reply:
Let's not do it for now.

Reply:
> I think a void* is better as the state type here. It's not a true variant, it can only hold one type for its entire lifetime.
> 

This is a good point. Will consider this approach once I revisit it in the future.